addons/mail/models/models.py

```diff
@@ -272,6 +273,28 @@ def _get_default_activity_view(self):
         templates = E.templates(activity_box)
         return E.activity(templates, string=self._description)
 
+    @api.model
+    def _sort_activities_by_record(self, activities, module=False):
+        """ Sort activities linked to records of the current model by record
+            and module. By default, all activities are sorted in the original module
+            of the current model. This method can be overriden in specific model
+            to change this behavior.
+
+        :param activities: mail.activity records whose res_model is the current model
+        :param module: if set, force the grouping of activities in `module` instead of
+                       the original module of the current model.
+        :return: a dictionary with modules as key containing a dictionary of
+                 activities sorted by records they are linked to
+        :rtype: dict(dict())
+                e.g. {'module 1': {record1: [activity1, ...], ...}, 'module2': ...}
+        """
+        if not module:
+            module = self._original_module
+        activities_by_rec_id_by_module = {module: defaultdict(lambda: self.env["mail.activity"])}
+        for activity in activities:
+            activities_by_rec_id_by_module[module][activity.res_id] += activity
+        return activities_by_rec_id_by_module
```

Comment:
with `module` as a variable inside the loop, it is not immediately clear that the result is a dict with a single key, you could rewrite it like this:

```suggestion
        return {module or self._original_module: {activity.rec_id: activity for activity in activities}}
```

or create the outer dictionary at the end and only iterate in the nested one.